JENKINS-47996: Don't record dependencies to self-generated artifacts

markusdlugi opened the pull request:
[JENKINS-47996](https://issues.jenkins-ci.org/browse/JENKINS-47996): Avoids recording dependencies to artifact which a pipeline has generated itself.

For this, I added an implementation of `equals()` and `hashCode()` to `MavenArtifact` so that we can just check whether `generatedArtifacts.contains(dependency.asMavenArtifact())`.

cyrille-leclerc commented:
Thanks @markusdlugi . I remember that you have a preference in the order in which we merge the PRs #110 and #109 . Shall I merge this one first?

cyrille-leclerc commented on jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/MavenSpyLogProcessor.java:
```diff
@@ -201,6 +202,39 @@ public String toString() {
+        
     }
 
     public static class MavenDependency extends MavenArtifact {
```
We will have to extract the domain objects soon :-)

I wanted to move to JAXB for the XML model and to be schema first (rather than code first) between the maven event spy and the jenkins plugin. 

I'm not sure how we could combine an Xsd2Java approach with these domain objects that would be more than POJOs and have domain methods. cc @alvarolobato

markusdlugi commented:
@cyrille-leclerc Yes, initially this one was supposed to be merged before #109, but they should be independent now after the changes in commit 2c3d654, so it doesn't really matter.

But since I had to change quite some things after the merge of #107, I'd first like to do some thorough testing of both PRs in our production environment. I'll let you know once I finished testing.

cyrille-leclerc commented:
> But since I had to change quite some things after the merge of #107, 
> I'd first like to do some thorough testing of both PRs in our production environment. I'll let you know once I finished testing.

What about a testing phase on the master branch then a "beta" version?

markusdlugi commented:
We can also do that, it's up to you :)
I have now also fixed the merge conflict in #109 and updated the listUpstreamJobs() to also take parent project dependencies into account. I haven't had the time to test it yet, but if you want, we can also merge it and test it on master.

markusdlugi commented:
@cyrille-leclerc I did a short test on our production Jenkins instance. It seems to work fine for the most part, except for multibranch pipelines. There, the branches will trigger each other, which is exactly the problem that was previously fixed by this PR. Seems like the parent pom changes have some effect on that. I'll try to find the issue tomorrow.

markusdlugi commented:
@cyrille-leclerc I found the issue - the problem was simply that projects which consumed a parent that they themselves defined would also lead to an entry in MAVEN_PARENT_PROJECT. Therefore, different branches of the same project would see the other branches as downstream projects and trigger them. I fixed this just like with normal dependencies by excluding parents which have been generated during the build.

After these changes it is working as intended in our environment. So as far as I'm concerned, this is mergeable.

cyrille-leclerc commented:
Thanks @markusdlugi !

cyrille-leclerc commented:
@markusdlugi thanks, I have to merge your PRs #109 and #110 for testing once we are confident on the fact that the version 3.1.0 released yesterday does not require an "emergency fix", I have in mind Friday.
We will then have few weeks to test your very valuable changes and we could release a GA version mid January.

markusdlugi commented:
@cyrille-leclerc Sure, let's do it like that :)
